testkit/jvm/src/main/scala/zio/test/ZSpecRunner.scala

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright 2019 John A. De Goes and the ZIO Contributors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package zio.test
+
+import zio.UIO
+
+trait ZSpecRunner {
+  def apply[R, E, L](spec: ZSpec[R, E, L]): UIO[ExecutedSpec[R, E, L]]
```

Comment:
This signature feels too restrictive – ZSpecRunner cannot provide an environment... How about `ZSpecRunner[R]` or maybe another entity that can override/combine Suite-specific Env with a Global Env?

Reply:
Why combine the runner with the provision of the environment? The provision can be deferred until much later.

Also: is your use case met by simply stitching together ordinary `Spec`, and having, e.g., an application-wide `ZIOTest`? You can compose Spec with `++`, for example, so you just need 1 per application / module.

Reply:
Also Spec's already have `provide` / `provideSome`

Reply:
> The provision can be deferred until much later.

Not in our usecase, e.g. we want to setup shared components/resources before all tests happen, then close them. So in this case it's `global provision -> run suites -> per-test provision -> run tests ; close per-test ; close global`

> Also: is your use case met by simply stitching together ordinary Spec, and having, e.g., an application-wide ZIOTest

Everything can be done by stitching all tests together! But it would be a bit of a nightmare to write and use such tests, esp. if we can just add a type parameter instead...

Reply:
If we want to allow R to participate globally, we'll need class tag on R.

I assume you want reflective discovery?

Reply:
Not sure if we do need a ClassTag:

- If you use non-reflective instantation: then the R is assembled in one place and checked statically – don't need a ClassTag.
- If you use reflective instantation – if ZTest includes a reference to the Runner that it prefers – it can be a direct instantiation, a label or some kind of a handle - it can be indexed by `R` and checked statically
- If you use reflective instantiation – AND if the decision is completely top-down, i.e. runner discovers all tests and tests do not choose their runner – a ClassTag would not really make it safe for the runner to manipulate the environment anyway, ClassTag contains only the java type which is not representative... Personally, in distage-testkit, we'd just process specific subclasses only and redirect non-distage tests to default runner, a ClassTag does not provide value in this case.